batik-gvt/src/main/java/org/apache/batik/gvt/font/AWTGlyphGeometryCache.java

```diff
@@ -39,160 +40,55 @@ public class AWTGlyphGeometryCache {
     protected static final int INITIAL_CAPACITY = 71;
 
     /**
-     * The underlying array
+     * The underlying map
      */
-    protected Entry[] table;
-
-    /**
-     * The number of entries
-     */
-    protected int count;
-
-    /**
-     * The reference queue.
-     */
-    protected ReferenceQueue referenceQueue = new ReferenceQueue();
+    ConcurrentHashMap<Character, SoftReference<Value>> cache;
 
     /**
      * Creates a new AWTGlyphGeometryCache.
      */
     public AWTGlyphGeometryCache() {
-        table = new Entry[INITIAL_CAPACITY];
+        this(INITIAL_CAPACITY);
     }
 
     /**
      * Creates a new AWTGlyphGeometryCache.
      * @param c The inital capacity.
      */
     public AWTGlyphGeometryCache(int c) {
-        table = new Entry[c];
+        cache = new ConcurrentHashMap<>(c);
     }
 
     /**
      * Returns the size of this table.
      */
     public int size() {
-        return count;
+        return cache.size();
     }
 
     /**
      * Gets the value of a variable
      * @return the value or null
      */
     public Value get(char c) {
-        int hash  = hashCode(c) & 0x7FFFFFFF;
-        int index = hash % table.length;
-
-        for (Entry e = table[index]; e != null; e = e.next) {
-            if ((e.hash == hash) && e.match(c)) {
-                return (Value)e.get();
-            }
-        }
-        return null;
+        return Optional.ofNullable(cache.get(c)).map(SoftReference::get).orElse(null);
     }
 
     /**
      * Sets a new value for the given variable
      * @return the old value or null
      */
     public Value put(char c, Value value) {
-        removeClearedEntries();
-
-        int hash  = hashCode(c) & 0x7FFFFFFF;
-        int index = hash % table.length;
-
-        Entry e = table[index];
-        if (e != null) {
-            if ((e.hash == hash) && e.match(c)) {
-                Object old = e.get();
-                table[index] = new Entry(hash, c, value, e.next);
-                return (Value)old;
-            }
-            Entry o = e;
-            e = e.next;
-            while (e != null) {
-                if ((e.hash == hash) && e.match(c)) {
-                    Object old = e.get();
-                    e = new Entry(hash, c, value, e.next);
-                    o.next = e;
-                    return (Value)old;
-                }
-
-                o = e;
-                e = e.next;
-            }
-        }
-
-        // The key is not in the hash table
-        int len = table.length;
-        if (count++ >= (len - (len >> 2))) {
-            // more than 75% loaded: grow
-            rehash();
-            index = hash % table.length;
-        }
-
-        table[index] = new Entry(hash, c, value, table[index]);
-        return null;
+        Value oldValue = get(c);
+        cache.put(c, new SoftReference<>(value));
+        return oldValue;
```

Comment:
With the helper method I mentioned in `get(char)`, this can be simplified to avoid two temporary Optionals and a map operation:

```
public Value put(char c, Value value) {
    return getIfSet(cache.put(c, new SoftReference<>(value));
}
```